crawl-ref/source/newgame.cc

```diff
@@ -233,38 +229,34 @@ static void _resolve_job(newgame_def& ng, const newgame_def& ng_choice)
         return;
 
     case JOB_VIABLE:
-        // If we have already picked a species, find a job that recommends it
-        if (is_starting_species(ng.species))
-        {
-            auto job_recommended_by_species =
-                filter_enum(job_type, [&](job_type job) {
-                    return species_recommends_job(job, ng.species);
-                });
-            // Note: we know the array has at least one element because all
-            // starting species must recommend at least one opposite job.
-            ng.job = job_recommended_by_species[
-                random2(job_recommended_by_species.size())];
-        }
-        else
-            ng.job = random_job();
-        return;
-
     case JOB_RANDOM:
-        // If we have already picked a species, find a job that allows it
+    {
+        job_type[] candidate_jobs;
         if (is_starting_species(ng.species))
         {
-            auto jobs_allowed_by_species =
-                filter_enum(job_type, [&](job_type job) {
-                    return character_is_allowed(ng.species, job);
-                });
-            if (jobs_allowed_by_species.size() == 0)
-                die("Selected species allows no job?!");
-            ng.job = jobs_allowed_by_species[
-                random2(jobs_allowed_by_species.size())];
+            switch (ng_choice.job)
+            {
+            case JOB_VIABLE:
+                // Note: we know the array has at least one element because all
+                // starting species must recommend at least one job.
+                candidate_jobs =
+                    filter_enum(job_type, [&](job_type job) {
+                        return species_recommends_job(job, ng.species);
+                    });
+            case JOB_RANDOM:
+                candidate_jobs =
+                    filter_enum(job_type, [&](job_type job) {
+                        return character_is_allowed(ng.species, job);
+                    });
+                if (candidate_species.size() == 0)
```

Comment:
this test should be outside the inner switch, so it tests for both VIABLE and RANDOM cases

Reply:
The messaging is different ("no recommended jobs" vs "no allowed jobs"), so I'll keep it inside the inner switch.